src/ophyd_async/core/_providers.py

```diff
@@ -17,41 +20,105 @@ class DirectoryInfo:
 
     :param root: Path of a root directory, relevant only for the file writer
     :param resource_dir: Directory into which files should be written, relative to root
-    :param prefix: Optional filename prefix to add to all files
-    :param suffix: Optional filename suffix to add to all files
+    :param filename: Base filename to use generated by FilenameProvider
+    :param create_dir_depth: Optional depth of directories to create if they do not exist
     """
 
     root: Path
     resource_dir: Path
-    prefix: Optional[str] = ""
-    suffix: Optional[str] = ""
+    filename: Optional[str] = ""
+    create_dir_depth: Optional[int] = 0
+
 
+class FilenameProvider(Protocol):
+
+    @abstractmethod
+    def __call__(self, prefix="", suffix="", device_name=None) -> str:
+        """Get a filename to use for output data"""
 
 class DirectoryProvider(Protocol):
+
+    _filename_provider: FilenameProvider
+    _prefix: Optional[str]
+    _suffix: Optional[str]
+
     @abstractmethod
-    def __call__(self) -> DirectoryInfo:
+    def __call__(self, device_name=None) -> DirectoryInfo:
         """Get the current directory to write files into"""
 
 
+class StaticFilenameProvider(FilenameProvider):
+    def __init__(self, filename):
+        self._static_filename = filename
+
+    def __call__(self, prefix="", suffix="", device_name=None) -> str:
+        return f'{prefix}{self._static_filename}{suffix}'
+
+
+class DeviceNameFilenameProvider(FilenameProvider):
+
+    def __call__(self, prefix="", suffix="", device_name=None) -> str:
+        return f'{prefix}{device_name}{suffix}'
```

Comment:
Is `WritesStreamAssets` a part of ophyd-async or dodal? I quite like using the device itself instead of a bunch of args here, then you can add any number of modifiers to the filename based on device specific attributes (even like naming files based on a signal reading perhaps?)

Reply:
Do we need anything from the Device apart from name?

Reply:
Just name <beamline>-<device.name>-<dataCollectionId>.h5

Reply:
So just passing `device_name` rather than `device` would be sufficient? Or do we think we will need anything else in the future from the object?

Reply:
If we're passing in a Device, all we can reliably expect is device.name and device.children AFAIK? device_name should be fine I guess? Each child that needs the DirectoryProvider will have the ability to call the NameProvider themselves